cranelift-codegen/src/binemit/relaxation.rs

```diff
@@ -134,6 +143,145 @@ pub fn relax_branches(func: &mut Function, isa: &dyn TargetIsa) -> CodegenResult
+        None => {
+            return false; // The instruction was either multi-target or not a branch.
+        }
+    };
+
+    // Look at the first instruction of the first branch's destination.
+    // If it is an unconditional branch, maybe the second jump can be bypassed.
+    let second_inst = func.layout.first_inst(first_dest).expect("Instructions");
+    if func.dfg[second_inst].opcode() != Opcode::Jump {
+        return false;
+    }
+
+    // Now we need to fix up first_inst's ebb parameters to match second_inst's,
+    // without changing the branch-specific arguments.
+    //
+    // The intermediary block is allowed to reference any SSA value that dominates it,
+    // but that SSA value may not necessarily also dominate the instruction that's
+    // being patched.
+
+    // Get the arguments and parameters passed by the first jump.
+    let num_fixed = func.dfg[first_inst]
+        .opcode()
+        .constraints()
+        .num_fixed_value_arguments();
+    let (first_args, first_params) = func.dfg[first_inst]
+        .arguments(&func.dfg.value_lists)
+        .split_at(num_fixed);
+
+    // Get the parameters passed by the second jump.
+    let num_fixed = func.dfg[second_inst]
+        .opcode()
+        .constraints()
+        .num_fixed_value_arguments();
+    let (_, second_params) = func.dfg[second_inst]
+        .arguments(&func.dfg.value_lists)
+        .split_at(num_fixed);
+    let mut second_params = second_params.to_vec(); // Clone for rewriting below.
+
+    // For each parameter passed by the second jump, if any of those parameters
+    // was a block parameter, rewrite it to refer to the value that the first jump
+    // passed in its parameters. Otherwise, make sure it dominates first_inst.
+    //
+    // For example: if we `ebb0: jump ebb1(v1)` to `ebb1(v2): jump ebb2 v2`,
+    // we want to rewrite the original jump to `jump ebb2(v1)`.
+    // Note that v2 does not dominate the first jump instruction.
+    let ebb_params: &[Value] = func.dfg.ebb_params(first_dest);
+    debug_assert!(ebb_params.len() == first_params.len());
+
+    for value in second_params.iter_mut() {
+        if let Some((n, _)) = ebb_params.iter().enumerate().find(|(_, &p)| p == *value) {
+            // This value was the Nth parameter passed to the second_inst's ebb.
+            // Rewrite it as the Nth parameter passed by first_inst.
+            *value = first_params[n];
+        } else {
+            // Check whether the parameter value dominates first_inst.
+            let dominates_branch = match func.dfg.value_def(*value) {
+                ValueDef::Result(inst, _) => domtree.dominates(inst, first_inst, &func.layout),
+                ValueDef::Param(ebb, _) => domtree.dominates(ebb, first_inst, &func.layout),
+            };
+
+            // The second_inst passes a parameter that doesn't dominate first_inst,
+            // so first_inst is unable to pass that parameter itself.
+            if !dominates_branch {
```

Comment:
I do not see how this condition would ever be possible if the `first_dest` is indeed just a `jump %second_dest` branch.
As any value which is not live from the ebb parameters of the first destination would necessarily comes from the predecessor of the first instruction.

I do not understand why this else branch is not a no-op. Do you have an example?
Otherwise I suggest to remove this else branch.

Reply:
I think you're right. I was trying to make the code safe against crazy CFGs. But you're correct: if this transformation were invalid, then the CFG would have originally been invalid, and we don't need to handle invalid CFGs.